addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,262 @@
+    notification_parameters = fields.Text('Notification parameters')
+
+    @api.constrains('scheduled_date')
+    def _check_scheduled_date(self):
+        if any(scheduled_message.scheduled_date < fields.Datetime().now() for scheduled_message in self):
+            raise ValidationError(_("A Scheduled Message cannot be scheduled in the past"))
+
+    # ------------------------------------------------------
+    # CRUD / ORM
+    # ------------------------------------------------------
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # check that messages can be posted on the models
+        for model in {vals['model'] for vals in vals_list}:
+            if not (model in self.pool and issubclass(self.pool[model], self.pool['mail.thread'])):
+                raise UserError(_("A message cannot be scheduled on a model that does not have a mail thread."))
+
+        Scheduled = self.browse()
+        for vals in vals_list:
+            # make sure user can post on the related records
+            Scheduled._check(vals)
+
+        scheduled_messages = super().create(vals_list)
+        # transfer attachments from composer to scheduled messages
+        for scheduled_message in scheduled_messages:
+            if attachments := scheduled_message.attachment_ids:
+                attachments.filtered(
+                    lambda a: a.res_model == 'mail.compose.message' and not a.res_id and a.create_uid.id == self.env.uid
+                ).write({
+                    'res_model': scheduled_message._name,
+                    'res_id': scheduled_message.id,
+                })
+        # schedule cron trigger
+        if scheduled_messages:
+            self.env.ref('mail.ir_cron_post_scheduled_message')._trigger_list(
+                set(scheduled_messages.mapped('scheduled_date'))
+            )
+        return scheduled_messages
+
+    @api.model
+    def _search(self, domain, offset=0, limit=None, order=None):
+        """ Override that add specific access rights to only get the ids of the messages
+        that are scheduled on the records on which the user has mail_post (or read) access
+        """
+        if self.env.is_superuser():
+            return super()._search(domain, offset, limit, order)
+
+        # Don't use the ORM to avoid cache pollution
+        query = super()._search(domain, offset, limit, order)
+        fnames_to_read = ['id', 'model', 'res_id']
+        rows = self.env.execute_query(query.select(
+            *[self._field_to_sql(self._table, fname) for fname in fnames_to_read],
+        ))
+
+        # group res_ids by model and determine accessible records
+        model_ids = defaultdict(set)
+        for __, model, res_id in rows:
+            model_ids[model].add(res_id)
+
+        allowed_ids = defaultdict(set)
+        for model, res_ids in model_ids.items():
+            records = self.env[model].browse(res_ids)
+            operation = records._mail_post_access
```

Comment:
In case of issue, would fallback on "write" is records has no atte "_mail_post_access" (people always end up creating weird stuff)

Reply:
aka ``operation = getattr(records, "_mail_post_access", "write")``